endpoint/endpoint_diagnoseComp/endpoint_diagnose.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "endpoint_diagnose.h"
+
+#include "legato.h"
+#include "le_gnss_interface.h"
+#include "le_mrc_interface.h"
+
+COMPONENT_INIT {
+
+  char time_str[30];
+  size_t copied_bytes = 0;
+  int32_t latitude = 0, longitude = 0, accuracy = 0;
+  uint32_t quality = 0;
+  le_gnss_SampleRef_t position_sample;
+  le_gnss_State_t gnss_state;
+  le_result_t result;
+
+  // GPS state checking and activation  
+  gnss_state = le_gnss_GetState();
+  if (gnss_state != LE_GNSS_STATE_ACTIVE) {
+    le_gnss_Start();
+    le_gnss_Enable();
+  }
+
+  while(1) {
+    // Signal strength
+    result = le_mrc_GetSignalQual(&quality);
+    if (result == LE_OK) {
+      LE_INFO("Signal strength: %d\n", quality);
+    } else {
+      LE_ERROR("Fail to get signal strength\n");
+    }
+
+    // GPS
+    position_sample = le_gnss_GetLastSampleRef();
+    result = le_gnss_GetLocation(position_sample, &latitude, &longitude, &accuracy);
+    if (result == LE_OK) {
+      LE_INFO("Latitude: %lf, Longitude: %lf, Horizontal accuracy: %d\n", (double)latitude / 1000000, (double)longitude / 1000000, accuracy);
+    } else if (result == LE_FAULT) {
+      LE_ERROR("Fail to get location data\n");
+    } else if (result == LE_OUT_OF_RANGE) {
+      LE_ERROR("At least one retrieved value is invalid (set to INT32_MAX)\n");
+      LE_INFO("Latitude: %lf, Longitude: %lf, Horizontal accuracy: %d\n", (double)latitude / 1000000, (double)longitude / 1000000, accuracy);
+    }
+
+    // Time
+    result = le_clk_GetLocalDateTimeString(LE_CLK_STRING_FORMAT_DATE_TIME, time_str, sizeof(time_str), &copied_bytes);
+    if (result == LE_OK) {
+      LE_INFO("Time: %s\n", time_str);
+    } else if (result == LE_OVERFLOW) {
+      LE_ERROR("Overflow of time string\n");
+    }
+
+    sleep(1);
```

Comment:
Is it possible to rewrite with [timerfd_create](https://man7.org/linux/man-pages/man2/timerfd_create.2.html)? Thus, we can avoid calling sleep and simplify the flow.

Reply:
I will rewrite it with the [timer API](https://docs.legato.io/latest/c_timer.html) offered by Legato framework.

Reply:
It sounds great. The downside to cycling sleep and running states rather is that interactive software will acquire a notable stutter if too little time is spent awake, too much time is spent sleeping, or a combination of both.

Reply:
Done.